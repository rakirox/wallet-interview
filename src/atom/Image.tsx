import React, {useState} from 'react';
import {StyleSheet, Image as BaseImage, ImageProps} from 'react-native';

export function Image(tProps: ImageProps): JSX.Element {
  const [loading, setLoading] = useState(true);
  return (
    <>
      <BaseImage
        {...tProps}
        style={[styles.image, tProps.style]}
        onLoadEnd={() => setLoading(false)}
      />
      ;
    </>
  );
}

const styles = StyleSheet.create({
  image: {},
});
