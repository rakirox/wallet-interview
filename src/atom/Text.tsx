import React from 'react';
import {StyleSheet, Text as BaseText, TextProps} from 'react-native';

export function Text(tProps: TextProps): JSX.Element {
  return <BaseText {...tProps} style={[styles.text, tProps.style]} />;
}

const styles = StyleSheet.create({
  text: {
    fontFamily: 'avenir',
  },
});
