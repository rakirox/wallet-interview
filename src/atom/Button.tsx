import React from 'react';
import {
  StyleSheet,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
} from 'react-native';
import {Text} from './Text';

type Props = {
  text: string;
  onPress: () => void;
  textStyle?: TextStyle;
} & TouchableOpacityProps;

export function Button({
  text,
  onPress,
  style,
  textStyle,
  ...tProps
}: Props): JSX.Element {
  return (
    <TouchableOpacity
      {...tProps}
      style={[styles.button, style]}
      onPress={onPress}>
      <Text style={[styles.text, textStyle, textStyle]}>{text}</Text>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#334FFA',
    borderRadius: 10,
  },
  text: {
    fontWeight: '800',
    fontSize: 16,
    lineHeight: 22,
    color: 'white',
    alignSelf: 'center',
    paddingVertical: 14,
  },
});
