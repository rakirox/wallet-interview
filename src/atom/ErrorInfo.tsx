import React, {PropsWithChildren} from 'react';
import {View, StyleSheet} from 'react-native';
import {Text} from './Text';

type Props = PropsWithChildren<{
  text: string;
}>;

export function ErrorInfo({text}: Props) {
  return (
    <View style={styles.container}>
      <Text style={styles.errorText}>Error encontrado: {text}</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
  },
  errorText: {
    color: 'red',
  },
});
