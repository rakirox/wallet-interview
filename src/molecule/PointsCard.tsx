import React, {PropsWithChildren} from 'react';
import {ActivityIndicator, StyleSheet, View} from 'react-native';
import {Text} from '../atom/Text';

type Props = PropsWithChildren<{
  title: string;
  points: number;
  loading: boolean;
}>;

export function PointsCard({title, points, loading}: Props): JSX.Element {
  return (
    <View style={styles.container}>
      <Text style={styles.title}>{title}</Text>
      {loading ? (
        <ActivityIndicator style={styles.activityIndicator} color="white" />
      ) : (
        <Text style={styles.points}>
          {`${points.toLocaleString('mxn', {
            minimumFractionDigits: 2,
          })} pts`}
        </Text>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  activityIndicator: {
    alignSelf: 'center',
    marginVertical: 30,
  },
  container: {
    backgroundColor: '#334FFA',
    padding: 20,
    shadowColor: 'rgba(0, 0, 0, 0.5)',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowRadius: 3.84,
    shadowOpacity: 1,
    elevation: 6,
    borderRadius: 20,
    minWidth: 280,
  },
  title: {
    fontWeight: '800',
    fontSize: 16,
    lineHeight: 22,
    color: '#ffffff',
    marginBottom: 7,
  },
  points: {
    fontWeight: '800',
    fontSize: 32,
    lineHeight: 44,
    color: '#ffffff',
    marginBottom: 50,
    alignSelf: 'center',
  },
});
