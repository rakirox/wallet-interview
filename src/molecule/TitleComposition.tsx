import React, {PropsWithChildren} from 'react';
import {StyleSheet} from 'react-native';
import {Text} from '../atom/Text';

type Props = PropsWithChildren<{
  title: string;
  subtitle: string;
}>;

export function TitleComposition({title, subtitle}: Props): JSX.Element {
  return (
    <>
      <Text style={styles.titleWelcome}>{title}</Text>
      <Text style={styles.subtitleWelcome}>{subtitle}</Text>
    </>
  );
}

const styles = StyleSheet.create({
  titleWelcome: {
    fontWeight: '800',
    fontSize: 20,
    lineHeight: 27.32,
    color: '#020202',
  },
  subtitleWelcome: {
    fontWeight: '400',
    fontSize: 16,
    lineHeight: 27.32,
    color: '#020202',
  },
});
