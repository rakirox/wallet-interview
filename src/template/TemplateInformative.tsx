import React, {PropsWithChildren} from 'react';
import {SafeAreaView, StyleSheet, View} from 'react-native';
import {Text} from '../atom/Text';

type TemplateProps = PropsWithChildren<{
  firstSectionTitle: string;
  scrollContentTitle: string;
  scrollContent: JSX.Element;
  firstSectionContent: JSX.Element;
  headerContent: JSX.Element;
  footContent: JSX.Element;
}>;

export function TemplateInformative({
  headerContent,
  scrollContentTitle,
  scrollContent,
  footContent,
  firstSectionTitle,
  firstSectionContent,
}: TemplateProps): JSX.Element {
  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.headerContainer}>{headerContent}</View>
      <Text style={styles.title}>{firstSectionTitle}</Text>
      <View style={styles.firsSectionContent}>{firstSectionContent}</View>
      <Text style={styles.title}>{scrollContentTitle}</Text>
      <View style={styles.scrollContext}>{scrollContent}</View>
      <View style={styles.footContent}>{footContent}</View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  footContent: {
    paddingHorizontal: 20,
  },
  container: {
    flex: 1,
  },
  headerContainer: {
    padding: 20,
  },
  title: {
    fontWeight: '800',
    color: '#9B9898',
    fontSize: 14,
    textTransform: 'uppercase',
    marginBottom: 20,
    marginLeft: 20,
  },
  firsSectionContent: {
    alignItems: 'center',
    marginBottom: 20,
  },
  scrollContext: {
    flex: 1,
  },
});
