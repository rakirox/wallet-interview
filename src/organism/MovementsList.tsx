import React, {PropsWithChildren} from 'react';
import {FlatList, StyleSheet} from 'react-native';
import {MovementItem} from '../molecule/MovementItem';
import {Movement} from '../types/Movement';

type Props = PropsWithChildren<{
  data: Movement[];
  onMovementPressed: (movement: Movement) => void;
}>;

export function MovementsList({data, onMovementPressed}: Props): JSX.Element {
  return (
    <FlatList
      style={styles.container}
      data={data}
      renderItem={({item}): JSX.Element => {
        return (
          <MovementItem
            movement={item}
            onPress={() => onMovementPressed(item)}
          />
        );
      }}
    />
  );
}

const styles = StyleSheet.create({
  container: {
    paddingVertical: 20,
    paddingHorizontal: 10,
  },
});
