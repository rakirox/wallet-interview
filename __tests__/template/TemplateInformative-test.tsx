import 'react-native';
import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import {TemplateInformative} from '../../src/template/TemplateInformative';
import {movements} from '../mock.json';
import {TouchableOpacity} from 'react-native';
import {TitleComposition} from '../../src/molecule/TitleComposition';
import {PointsCard} from '../../src/molecule/PointsCard';
import {MovementsList} from '../../src/organism/MovementsList';
import {Button} from '../../src/atom/Button';

describe('TemplateInformative', () => {
  it('render correctly', () => {
    expect(
      renderer.create(
        <TemplateInformative
          headerContent={
            <TitleComposition
              title="Bienvenido de vuelta!"
              subtitle="Ruben Rodriguez"
            />
          }
          firstSectionTitle="Tus Puntos"
          firstSectionContent={
            <PointsCard
              title="Diciembre"
              points={movements.reduce(
                (acc, item) =>
                  acc + (item.isRedemption ? -item.points : item.points),
                0,
              )}
              loading={false}
            />
          }
          scrollContentTitle="tus movimientos"
          scrollContent={
            <MovementsList data={movements} onMovementPressed={jest.fn()} />
          }
          footContent={<Button text="Todos" onPress={jest.fn()} />}
        />,
      ),
    ).toMatchSnapshot();
  });
});
