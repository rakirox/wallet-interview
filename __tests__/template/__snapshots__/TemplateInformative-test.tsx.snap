// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateInformative render correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "padding": 20,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "avenir",
          },
          {
            "color": "#020202",
            "fontSize": 20,
            "fontWeight": "800",
            "lineHeight": 27.32,
          },
        ]
      }
    >
      Bienvenido de vuelta!
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "avenir",
          },
          {
            "color": "#020202",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 27.32,
          },
        ]
      }
    >
      Ruben Rodriguez
    </Text>
  </View>
  <Text
    style={
      [
        {
          "fontFamily": "avenir",
        },
        {
          "color": "#9B9898",
          "fontSize": 14,
          "fontWeight": "800",
          "marginBottom": 20,
          "marginLeft": 20,
          "textTransform": "uppercase",
        },
      ]
    }
  >
    Tus Puntos
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "marginBottom": 20,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#334FFA",
          "borderRadius": 20,
          "elevation": 6,
          "minWidth": 280,
          "padding": 20,
          "shadowColor": "rgba(0, 0, 0, 0.5)",
          "shadowOffset": {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 3.84,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "avenir",
            },
            {
              "color": "#ffffff",
              "fontSize": 16,
              "fontWeight": "800",
              "lineHeight": 22,
              "marginBottom": 7,
            },
          ]
        }
      >
        Diciembre
      </Text>
      <Text
        style={
          [
            {
              "fontFamily": "avenir",
            },
            {
              "alignSelf": "center",
              "color": "#ffffff",
              "fontSize": 32,
              "fontWeight": "800",
              "lineHeight": 44,
              "marginBottom": 50,
            },
          ]
        }
      >
        0.00 pts
      </Text>
    </View>
  </View>
  <Text
    style={
      [
        {
          "fontFamily": "avenir",
        },
        {
          "color": "#9B9898",
          "fontSize": 14,
          "fontWeight": "800",
          "marginBottom": 20,
          "marginLeft": 20,
          "textTransform": "uppercase",
        },
      ]
    }
  >
    tus movimientos
  </Text>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      data={
        [
          {
            "createdAt": "2022-12-09T10:20:00.909Z",
            "id": "3",
            "image": "https://loremflickr.com/640/480/technics",
            "isRedemption": false,
            "points": 42416,
            "product": "Fantastic Granite Bacon",
          },
          {
            "createdAt": "2022-12-09T10:20:00.909Z",
            "id": "2",
            "image": "https://loremflickr.com/640/480/technics",
            "isRedemption": true,
            "points": 42416,
            "product": "Fantastic Granite Bacon 2",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      style={
        {
          "paddingHorizontal": 10,
          "paddingVertical": 20,
        }
      }
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "opacity": 1,
                "paddingBottom": 20,
                "paddingHorizontal": 20,
              }
            }
          >
            <View
              style={
                {
                  "borderRadius": 10,
                  "height": 55,
                  "marginRight": 11,
                  "overflow": "hidden",
                  "width": 55,
                }
              }
            >
              <Image
                onLoadEnd={[Function]}
                source={
                  {
                    "uri": "https://loremflickr.com/640/480/technics",
                  }
                }
                style={
                  [
                    {},
                    {
                      "flex": 1,
                    },
                  ]
                }
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              >
                <ActivityIndicator
                  color="rgba(0,0,0,0.5)"
                />
              </View>
            </View>
            <View
              style={
                {
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "avenir",
                    },
                    {
                      "fontSize": 14,
                      "fontWeight": "800",
                      "marginBottom": 7,
                    },
                  ]
                }
              >
                Fantastic Granite Bacon
              </Text>
              <Text
                style={
                  [
                    {
                      "fontFamily": "avenir",
                    },
                    {
                      "fontSize": 12,
                      "fontWeight": "400",
                    },
                  ]
                }
              >
                09 de December, 2022
              </Text>
            </View>
            <Text
              style={
                [
                  {
                    "fontFamily": "avenir",
                  },
                  {
                    "fontSize": 16,
                    "fontWeight": "800",
                    "lineHeight": 22,
                    "marginRight": 23,
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "avenir",
                    },
                    {
                      "color": "#00B833",
                      "fontSize": 16,
                      "fontStyle": "normal",
                      "fontWeight": "800",
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                +
              </Text>
              42416
            </Text>
            <View
              style={{}}
            >
              <Image
                onLoadEnd={[Function]}
                source={
                  {
                    "testUri": "../../../src/assets/images/subtract.png",
                  }
                }
                style={
                  [
                    {},
                    {
                      "height": 10,
                      "width": 10,
                    },
                  ]
                }
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              >
                <ActivityIndicator
                  color="rgba(0,0,0,0.5)"
                />
              </View>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "opacity": 1,
                "paddingBottom": 20,
                "paddingHorizontal": 20,
              }
            }
          >
            <View
              style={
                {
                  "borderRadius": 10,
                  "height": 55,
                  "marginRight": 11,
                  "overflow": "hidden",
                  "width": 55,
                }
              }
            >
              <Image
                onLoadEnd={[Function]}
                source={
                  {
                    "uri": "https://loremflickr.com/640/480/technics",
                  }
                }
                style={
                  [
                    {},
                    {
                      "flex": 1,
                    },
                  ]
                }
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              >
                <ActivityIndicator
                  color="rgba(0,0,0,0.5)"
                />
              </View>
            </View>
            <View
              style={
                {
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "avenir",
                    },
                    {
                      "fontSize": 14,
                      "fontWeight": "800",
                      "marginBottom": 7,
                    },
                  ]
                }
              >
                Fantastic Granite Bacon 2
              </Text>
              <Text
                style={
                  [
                    {
                      "fontFamily": "avenir",
                    },
                    {
                      "fontSize": 12,
                      "fontWeight": "400",
                    },
                  ]
                }
              >
                09 de December, 2022
              </Text>
            </View>
            <Text
              style={
                [
                  {
                    "fontFamily": "avenir",
                  },
                  {
                    "fontSize": 16,
                    "fontWeight": "800",
                    "lineHeight": 22,
                    "marginRight": 23,
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "avenir",
                    },
                    {
                      "color": "#FF0000",
                      "fontSize": 16,
                      "fontStyle": "normal",
                      "fontWeight": "800",
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                -
              </Text>
              42416
            </Text>
            <View
              style={{}}
            >
              <Image
                onLoadEnd={[Function]}
                source={
                  {
                    "testUri": "../../../src/assets/images/subtract.png",
                  }
                }
                style={
                  [
                    {},
                    {
                      "height": 10,
                      "width": 10,
                    },
                  ]
                }
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              >
                <ActivityIndicator
                  color="rgba(0,0,0,0.5)"
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      {
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#334FFA",
          "borderRadius": 10,
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "avenir",
            },
            [
              {
                "alignSelf": "center",
                "color": "white",
                "fontSize": 16,
                "fontWeight": "800",
                "lineHeight": 22,
                "paddingVertical": 14,
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        Todos
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
